### PR TITLE
Ubuntu2004 image support on libvirt and aws

### PR DESCRIPTION
**Pull request by rjmateus**
## What does this PR change?

Issue: https://github.com/SUSE/spacewalk/issues/11489
Draft PR to add ubuntu 20.04 image support for libvirt and aws.

After clients tools get published on Uyuni, we need to update the repositories in use on cloud-init user data.

### CODE REVIEW

File: backend_modules/libvirt/host/user_data.yaml [hunk 1]
```diff
@@ -192,7 +192,24 @@ runcmd:
...
+%{ endif }
+
+%{ if image == "ubuntu2004o"}
+
+apt:
+  sources:
+    tools_pool_repo:
+      # TODO after release, update this repository to use UYUNI instead
+      #source: "deb http://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/Ubuntu2004-Uyuni-Client-Tools/xUbuntu_20.04 /"
+      source: "deb [trusted=yes] http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/Head:/Ubuntu20.04-SUSE-Manager-Tools/xUbuntu_20.04 /"
```
Comment (juliogonzalez) [inline backend_modules/libvirt/host/user_data.yaml:212]: If you are using Head, use the Master Client Tools for Uyuni instead: https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu2004-Uyuni-Client-Tools/xUbuntu_20.04/

And then when we have the Stable client tools, you just need `Master` to `Stable`

Same comment about `[trusted=yes]`

File: backend_modules/aws/host/user_data.yaml [hunk 2]
```diff
@@ -110,6 +110,19 @@ runcmd:
...
 %{ endif }
+%{ if image == "ubuntu2004" }
+apt:
+  sources:
+    tools_pool_repo:
+      source: "deb [trusted=yes] http://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/Ubuntu2004-Uyuni-Client-Tools/xUbuntu_20.04 /"
+
+runcmd:
+  # HACK: cloud-init in Ubuntu does not take care of the following
+  - echo "PermitRootLogin yes" >> /etc/ssh/sshd_config
```
[the remark this answers is not in the document]
Comment (rjmateus) [reply backend_modules/aws/host/user_data.yaml:121]: The test suite will try to onboard all ssh clients using "root" user, so we need to enable it in all images.

File: backend_modules/libvirt/host/user_data.yaml [hunk 3]
```diff
@@ -192,7 +192,24 @@ runcmd:
 packages: ["salt-minion", "avahi-daemon", "qemu-guest-agent"]
 %{ endif }
 
-%{ if image == "ubuntu1804o" }
+%{ if image == "ubuntu1804o"}
```
[the remark this answers is not in the document]
Comment (rjmateus) [reply backend_modules/libvirt/host/user_data.yaml:195]: I will add it again

File: backend_modules/libvirt/host/user_data.yaml [hunk 4]
```diff
@@ -192,7 +192,24 @@ runcmd:
...
+
+runcmd:
+  # HACK: cloud-init in Ubuntu does not take care of the following
+  - echo "PermitRootLogin yes" >> /etc/ssh/sshd_config
```
[the remark this answers is not in the document]
Comment (rjmateus) [reply backend_modules/libvirt/host/user_data.yaml:199]: some motive

File: backend_modules/libvirt/host/user_data.yaml (hunk 1 above)
Comment (rjmateus) [reply backend_modules/libvirt/host/user_data.yaml:212]: oik, good point. I will add it

File: backend_modules/aws/host/user_data.yaml [hunk 5]
```diff
@@ -110,6 +110,19 @@ runcmd:
   # HACK: needed because onboard traditional clients
   - "/usr/bin/update-ca-trust force-enable"
 %{ endif }
+%{ if image == "ubuntu2004" }
+apt:
+  sources:
+    tools_pool_repo:
+      source: "deb [trusted=yes] http://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Stable:/Ubuntu2004-Uyuni-Client-Tools/xUbuntu_20.04 /"
```
[the remark this answers is not in the document]
Comment (rjmateus) [reply backend_modules/aws/host/user_data.yaml:117]: I will add it again. Thanks

**Comment by rjmateus** [comment]
@juliogonzalez updated. I didn't test it yet. Is running now on my machine

File: backend_modules/aws/host/user_data.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (juliogonzalez) [reply backend_modules/aws/host/user_data.yaml:144]: Ok, then sumaform is right (for now), but the testsuite is not (customers/users don't do that!)

@Bischoff you already have this on your radar, right?

**Comment by rjmateus** [comment]
@juliogonzalez When trying to use teh ibs devel repo I'm getting this error: 
```
module.ubuntu_2004_official.module.minion.module.host.null_resource.provisioning[0] (remote-exec):      Comment: Failed to configure repo 'deb http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/Head:/Ubuntu20.04-SUSE-Manager-Tools/xUbuntu_20.04 /': Error: [Errno -2] Name or service not known reading http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/Head:/Ubuntu20.04-SUSE-Manager-Tools/xUbuntu_20.04/Release.key
```
Do you have a clue about what could go wrong?

**Comment by juliogonzalez** [comment]
> Do you have a clue about what could go wrong?

Either you are not connected to the VPN (as IBS is internal), either you have DNS problems. If it's the second case, how to fix it will depend on your setup. If the VPN DNS are configured at /etc/resolv.conf after connecting, make sure libvirt is using it, as well as the instances.

**Comment by rjmateus** [comment]
Thanks @juliogonzalez I re-run the test locally and it now worked. In the previous run I temember I had the VPN on, but maybe I have done something else wrong. I will merge now

**Comment by rjmateus** [comment]
adding @Bischoff so he can be aware of this PR.

File: backend_modules/aws/host/user_data.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Bischoff) [reply backend_modules/aws/host/user_data.yaml:144]: @juliogonzalez @ricardomateus I knew about the issue, but it was not formalized. Here is a new card to track it:

https://github.com/SUSE/spacewalk/issues/11505

**Pull request merged**